vm/cuttlefish/cuttlefish.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+// Package cuttlefish allows to use Cuttlefish Android emulators hosted on Google Compute Engine
+// (GCE) virtual machines as VMs. It is assumed that syz-manager also runs on GCE as VMs are
+// created in the current project/zone.
+//
+// See https://cloud.google.com/compute/docs for details.
+// In particular, how to build GCE-compatible images:
+// https://cloud.google.com/compute/docs/tutorials/building-images
+package cuttlefish
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/syzkaller/pkg/config"
+	"github.com/google/syzkaller/pkg/gce"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/vm/vmimpl"
+
+	gcevm "github.com/google/syzkaller/vm/gce"
+)
+
+func init() {
+	vmimpl.Register("cuttlefish", ctor, true)
+}
+
+type Config struct {
+	Count       int    `json:"count"`        // number of VMs to use
+	MachineType string `json:"machine_type"` // GCE machine type (e.g. "n1-standard-4")
+	GCEImage    string `json:"gce_image"`    // pre-created GCE image to use
+	Preemptible bool   `json:"preemptible"`  // use preemptible VMs if available (defaults to true)
+}
+
+type Pool struct {
+	cfg     *Config
+	env     *vmimpl.Env
+	gcePool *gcevm.Pool
+}
+
+type instance struct {
+	debug   bool
+	gceInst vmimpl.Instance
+}
+
+func ctor(env *vmimpl.Env) (vmimpl.Pool, error) {
+	log.Logf(0, "cuttlefish.ctor()")
+	if env.Name == "" {
```

Comment:
Humm... here we circumvent lots of vm/gce functionality, but also duplicate the same logic.
Can't we do something like:
```
gcePool, err := vmimpl.Types["gce"].Ctor(env)
```
?
Then we don't need own duplicate Config, config checking logic, nor change vm/gce to expose all fields as public.

Reply:
Done.

Originally I was anticipating that we may eventually have additional config items for Cuttlefish beyond what GCE has, but I suppose we can address that if/when we need it.

Reply:
I think it should be possible to extend the config with additional fields and still reuse the gce constructor.
In vm/cuttlefish we can deserialize the passed config into an extended struct:
```
type Config struct {
  gce.Config
  ... other fields ...
}
```
Then we will need to somehow remove "... other fields ..." before passing them to vm/gce ctor (so that it does not fail on unknown fields). That should be doable.